plugins/outputs/azure_data_explorer/azure_data_explorer.go

```diff
@@ -0,0 +1,196 @@
+  endpoint_url = ""
+  
+  ## The name of the database in Azure Data Explorer where the ingestion will happen
+  database = ""
+
+  ## The client ID of the Service Principal in Azure that has ingestion rights to the Azure Data Exploer Cluster
+  client_id = ""
+
+  ## The client secret of the Service Principal in Azure that has ingestion rights to the Azure Data Exploer Cluster
+  client_secret = ""
+
+  ## The tenant ID of the Azure Subsciption in which the Service Principal belongs to
+  tenant_id = ""
+
+  ## Timeout for Azure Data Explorer operations
+  # timeout = "15s"
+`
+}
+
+func (adx *AzureDataExplorer) Connect() error {
+	authorizer := kusto.Authorization{
+		Config: auth.NewClientCredentialsConfig(adx.ClientID, adx.ClientSecret, adx.TenantID),
+	}
+
+	client, err := kusto.New(adx.Endpoint, authorizer)
+
+	if err != nil {
+		return err
+	}
+	adx.client = client
+	adx.ingesters = make(map[string]localIngestor)
+	adx.createIngestor = createRealIngestor
+
+	return nil
+}
+
+func (adx *AzureDataExplorer) Close() error {
+	adx.client = nil
+	adx.ingesters = nil
+
+	return nil
+}
+
+func (adx *AzureDataExplorer) Write(metrics []telegraf.Metric) error {
+	metricsPerNamespace := make(map[string][]byte)
+	// Group metrics by name and serialize them
+	for _, m := range metrics {
+		namespace := m.Name()
+		metricInBytes, err := adx.serializer.Serialize(m)
+		if err != nil {
+			return err
+		}
+		if existingBytes, ok := metricsPerNamespace[namespace]; ok {
+			metricsPerNamespace[namespace] = append(existingBytes, metricInBytes...)
+		} else {
+			metricsPerNamespace[namespace] = metricInBytes
+		}
+	}
+	ctx := context.Background()
+	ctx, cancel := context.WithTimeout(ctx, time.Duration(adx.Timeout))
+	defer cancel()
+
+	// Push the metrics namespace-wise
+	format := ingest.FileFormat(ingest.JSON)
+	for namespace, mPerNamespace := range metricsPerNamespace {
+		reader := bytes.NewReader(mPerNamespace)
+		//create a table for the namespace if it doesn't exist
+		if adx.ingesters[namespace] == nil {
+			if err := createAzureDataExplorerTableForNamespace(ctx, adx.client, adx.Database, namespace); err != nil {
+				return fmt.Errorf("creating table for %q failed: %v", namespace, err)
+			}
+			//create a new ingestor client for the namespace
+			ingestor, err := adx.createIngestor(adx.client, adx.Database, namespace)
+			if err != nil {
+				return fmt.Errorf("creating ingestor for %q failed: %v", namespace, err)
+			}
+			adx.ingesters[namespace] = ingestor
+		}
+
+		mapping := ingest.IngestionMappingRef(fmt.Sprintf("%s_mapping", namespace), ingest.JSON)
+		if _, err := adx.ingesters[namespace].FromReader(ctx, reader, format, mapping); err != nil {
+			adx.Log.Errorf("sending ingestion request to Azure Data Explorer for metric %q failed: %v", namespace, err)
+		}
+	}
+	return nil
+}
+
+func createAzureDataExplorerTableForNamespace(ctx context.Context, client localClient, database string, tableName string) error {
```

Comment:
Would it make sense to make this a member of `(adx *AzureDataExplorer)`? This way you don't need to pass `client` and `database` as parameters.

Reply:
I usually try to pass thin parameters to my functions to give flexibility for testing and make the dependencies clear. This is part of this habit.
Nonetheless, I have changed the code according to the suggestion :).